Add --ec2-user-data option to launch command

SoyeonBaek-dev opened the pull request:
This PR adds `--ec2-user-data` option to launch command.
This flag can help users create `UserData` defined instances. 
`--ec2-user-data` requires the path of `UserData` script file, and it could support long and complex script under 16KB.[1]

In case of increasing the number of slave instances, new slave instances are going to have the same `userdata` from the master instance.

I tested this PR by deploying cluster to AWS.


This PR is related to [#142](https://github.com/nchammas/flintrock/pull/142).
[1]: [Configuring Instances with User Data](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html)

SoyeonBaek-dev commented:
I found the error when requesting the spot instances and fixed it. 
I'm testing now, and let you know when I'm done.

nchammas commented on flintrock/flintrock.py:
```diff
@@ -226,6 +226,8 @@ def cli(cli_context, config, provider):
 @click.option('--ec2-ebs-optimized/--no-ec2-ebs-optimized', default=False)
 @click.option('--ec2-instance-initiated-shutdown-behavior', default='stop',
               type=click.Choice(['stop', 'terminate']))
+@click.option('--ec2-user-data', default='',
```
Would it be appropriate to use Click's `Path` type here? I think it would help us eliminate the need for `_get_user_data()`.

References:
* http://click.pocoo.org/6/api/#click.Path
* http://click.pocoo.org/6/arguments/#file-path-arguments

Example from Flintrock:
* https://github.com/nchammas/flintrock/blob/7e811104184e1d45e02bf0f4056fbc2d53ba2746/flintrock/flintrock.py#L206-L208

SoyeonBaek-dev replied:
That's great. I missed that point. 
I would change it to Click's `Path` type and remove `_get_user_data()`.

SoyeonBaek-dev replied:
I changed `--ec2-user-data` option type to Click's `File` type, and remove `_get_user_data()`.  
I think Click's `File` type is more appropriate than Click's `Path` type since user-data script should be opened and read. 
Could you give me a comment for this?

nchammas replied:
`File` sounds good to me!

nchammas commented on flintrock/flintrock.py:
```diff
@@ -226,6 +226,8 @@ def cli(cli_context, config, provider):
 @click.option('--ec2-instance-initiated-shutdown-behavior', default='stop',
               type=click.Choice(['stop', 'terminate']))
+@click.option('--ec2-user-data', default='',
+              help="Path to userdata script file for accessing nodes.")
```
I don't understand what you mean here by "for accessing nodes". Perhaps we could just phrase this as follows?

> Path to EC2 user data script that will run on instance launch.

SoyeonBaek-dev replied:
Oh, I didn't fix the typo. Thanks for checking it.

nchammas commented:
Ah, I just saw your most recent comment, @SoyeonBaek-dev. OK cool.

SoyeonBaek-dev commented on flintrock/ec2.py:
```diff
@@ -3,6 +3,8 @@
 import sys
 import time
 import urllib.request
+import base64
```
Yes.  As you said I didn't use `base64` module in this commit, but I need it in next commit. 

`ec2.create_instances()` supports automatically base64 encoded `UserData`, however, 
`ec2.meta.client.request_spot_instances()` doesn't seem to support that.

 I think I should give base64 encoded string `user_data` to `ec2.meta.client.request_spot_instances()` `UserData` parameter. 

This is what I'm dealing with in above 'error when requesting the spot instances' problem.

SoyeonBaek-dev commented:
I fixed the bug about requesting spot instances with user-data.
Also, I changed the launch `--ec2-user-data` option type and help message. 

Please review, and feel free to give me any comments.

nchammas commented:
Updated PR looks good to me. Going to test this out now.

nchammas commented:
I'm seeing this when I try to add slaves to a regular, on-demand cluster:

```
$ flintrock add-slaves nick --num-slaves 1
Traceback (most recent call last):
  File ".../venv/bin/flintrock", line 9, in <module>
    load_entry_point('Flintrock==0.7.0.dev0', 'console_scripts', 'flintrock')()
  File ".../venv/lib/python3.5/site-packages/flintrock/flintrock.py", line 1039, in main
    cli(obj={})
  File ".../venv/lib/python3.5/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File ".../venv/lib/python3.5/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File ".../venv/lib/python3.5/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File ".../venv/lib/python3.5/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File ".../venv/lib/python3.5/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File ".../venv/lib/python3.5/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File ".../venv/lib/python3.5/site-packages/flintrock/flintrock.py", line 661, in add_slaves
    **provider_options)
  File ".../venv/lib/python3.5/site-packages/flintrock/ec2.py", line 48, in wrapper
    res = func(*args, **kwargs)
  File ".../venv/lib/python3.5/site-packages/flintrock/ec2.py", line 283, in add_slaves
    user_data=user_data)
  File ".../venv/lib/python3.5/site-packages/flintrock/ec2.py", line 668, in _create_instances
    user_data = user_data.read()
AttributeError: 'str' object has no attribute 'read'
```

I think the issue is that `_create_instances()` should probably be passed the file contents rather than the file handle during `launch`, which would make things consistent with `add-slaves`.

nchammas commented on flintrock/ec2.py:
```diff
@@ -647,14 +659,17 @@ def _create_instances(
 
     cluster_instances = []
     spot_requests = []
+    user_data = user_data.read()
```
To follow-on to my comment on the main PR, we should remove this line and have the methods calling `_create_instances()` take care of reading the file contents where appropriate.

SoyeonBaek-dev replied:
I removed that 668 line. 
Instead of that, reading the user-data file contents is located in the `launch` before calling `_create_instances()` if `--ec2-user-data` is given.

nchammas commented:
@SoyeonBaek-dev - When you fix this issue, just confirm to me that you've successfully tested the following scenarios and I will merge this in:

1. Launching a regular cluster.
2. Adding slaves to a regular cluster.
3. Launching a spot cluster.
4. Adding slaves to a spot cluster.

SoyeonBaek-dev commented:
I should have tested all the test in the flintrock, and now I passed all the test except `test_launch_with_bad_ami` test(this test should be fixed).
Also, I checked all the scenarios you mentioned above. 

Please check and test once again.

nchammas commented:
Retested this and it all works as expected now. Thank you @SoyeonBaek-dev!
